server/service/middleware/ratelimit/ratelimit.go

```diff
@@ -47,6 +48,53 @@ func (m *Middleware) Limit(keyName string, quota throttled.RateQuota) endpoint.M
 	}
 }
 
+// ErrorMiddleware is a rate limiter that performs limits only when there is an error in the request
+type ErrorMiddleware struct {
+	store throttled.GCRAStore
+}
+
+// NewErrorMiddleware creates a new instance of ErrorMiddleware
+func NewErrorMiddleware(store throttled.GCRAStore) *ErrorMiddleware {
+	if store == nil {
+		panic("nil store")
+	}
+
+	return &ErrorMiddleware{store: store}
+}
+
+// Limit returns a new middleware function enforcing the provided quota only when errors occur in the next middleware
+func (m *ErrorMiddleware) Limit(keyName string, quota throttled.RateQuota) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		limiter, err := throttled.NewGCRARateLimiter(m.store, quota)
+		if err != nil {
+			panic(err)
+		}
+
+		return func(ctx context.Context, req interface{}) (response interface{}, err error) {
+			xForwardedFor, _ := ctx.Value(kithttp.ContextKeyRequestXForwardedFor).(string)
+			ipKeyName := fmt.Sprintf("%s-%s", keyName, xForwardedFor)
```

Comment:
should `ContextKeyRequestRemoteAddr` also be part of the key? in case `X-Forwarded-For` is not present